symfony/plugins/orangehrmMaintenancePlugin/lib/accessStrategies/AccessStrategy.php

```diff
@@ -0,0 +1,159 @@
+ * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along with this program;
+ * if not, write to the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301, USA
+ */
+
+/**
+ * Class AccessStrategy
+ */
+abstract class AccessStrategy
+{
+
+    protected $parameters = array();
+    protected $entityClassName;
+    protected $matchByArray = array();
+    protected $matchingCriteria;
+    protected $maintenanceService;
+    protected $getRealValueClass;
+
+    /**
+     * AccessStrategy constructor.
+     * @param $entityClassName
+     * @param $infoArray
+     */
+    public function __construct($entityClassName, $infoArray)
+    {
+        $this->setEntityClassName($entityClassName);
+        if (isset($infoArray['match_by'])) {
+            $this->setMatchBy($infoArray['match_by']);
+        }
+        if (isset($infoArray['parameters'])) {
+            $this->setParameters($infoArray['parameters']);
+        }
+    }
+
+    /**
+     * @param $employeeNumber
+     * @return mixed
+     */
+    public abstract function access($employeeNumber);
+
+    /**
+     * @return array
+     */
+    public function getParameters()
+    {
+        return $this->parameters;
+    }
+
+    /**
+     * @param $parametersArray
+     */
+    public function setParameters($parametersArray)
+    {
+        $this->parameters = $parametersArray;
+    }
+
+    /**
+     * @return mixed
+     */
+    public function getEntityClassName()
+    {
+        return $this->entityClassName;
+    }
+
+    /**
+     * @param $entityClassName
+     */
+    public function setEntityClassName($entityClassName)
+    {
+        $this->entityClassName = $entityClassName;
+    }
+
+
+    /**
+     * @return array
+     */
+    public function getMatchBy()
+    {
+        return $this->matchByArray;
+    }
+
+    /**
+     * @param $matchByArray
+     */
+    public function setMatchBy($matchByArray)
+    {
+        $this->matchByArray = $matchByArray;
+    }
+
+    /**
+     * @return MaintenanceService
+     */
+    public function getMaintenanceService()
+    {
+        if (!isset($this->maintenanceService)) {
+            $this->maintenanceService = new MaintenanceService();
+        }
+        return $this->maintenanceService;
+    }
+
+    /**
+     * @param MaintenanceService $maintenanceService
+     */
+    public function setMaintenanceService($maintenanceService)
+    {
+        $this->maintenanceService = $maintenanceService;
+    }
+
+    /**
+     * @param $employeeNumber
+     * @return array
+     */
+    public function getMatchByValues($employeeNumber)
+    {
+        $matchValueArray = array();
+
+        $matchValueArray[$this->getMatchBy()[0]['match']] = $employeeNumber;
```

Comment:
What happens if $this->getMatchBy()[0] is null

Reply:
this is hard coded in a yml file and it is reading value from that. and for all table this is a compulsory field declare in yml file.

Reply:
It's better to handle the null case and throw an exception.. otherwise, this code only works with that .yml file ( high coupling )